contents/testing/writing-testable-code.md

```diff
@@ -0,0 +1,110 @@
+# Writing Testable Code
+
+Below is a collection of 10 rules that you can follow to write testable code.
+
+## Don't mix object construction with application logic
+
+Because when you are writing tests, you mostly want to isolate some part of the application.
+
+In order to instantiate the class in isolation you have to make sure that the class itself does not instantiate other objects.
+
+So the first rule for writing testable code is do not mix object construction with application logic, instead, you should have two kinds of classes.
+
+First, the **factories**, where all `new` operators reside in, are in charge of constructing objects, and nothing else.
+
+The other, the **application logic classes**, are devoid of `new` operators. They simply ask for them. Thus it makes easier to test the application logic by replacing the real classes for test doubles.
+
+## Don't look for things. Ask for things.
+
+Aka Dependency Injection / Law of Demeter.
+
+The [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) can be highlighted as:
+
+* Only talk to your **immediate** friends.
+
+> If you are an Authenticator class don't pass in a path of the configuration information which you read inside the constructor to configure yourself, just ask for the configuration object and let some other class worry about reading the object from the disk.
```

Comment:
Ok! Say there is a `Config` class. And you can get a `Config` by reading a configuration file stored in disk. Then there is a class A needs a `Config`, do you inject a configuration file path, or do you pass in a `Config` directly?

Why injecting a file path is not good? Think about what you need to at testing, you will need to write some configuration to a file first, then pass in the file path to A, and let A read a `Config` itself. Versus you can directly create a `Config` object on the fly and pass it to A.

In practice, the reason many people compain Java is because it's too verbose. You do `a.getB().getC().getD().doSomething()`. This rule forbids this. You should never use more than one dot.